drivers/sps30/sps30.c

```diff
@@ -0,0 +1,287 @@
+ *
+ * @details  The CRC contained in the `crcd_data` is checked and the payload is
+ *           copied to `data`.
+ *           The format of `data` will be just the pure payload:
+ *           { data[0], data[1], data[2], ..., data[n]}
+ *           The format of `crcd_data` is same as it is read from the SPS30.
+ *           It consists of payload byte-pairs followed by a single CRC byte:
+ *           { data[0], data[1], csum[0],.., data[n-1], data[n], csum[n-1] }
+ *
+ * @pre          sizeof(crcd_data) must be equal to 1.5 * len
+ *
+ * @param[out]   data        Destination buffer for just the raw payload
+ * @param[in]    len         Size of the 'data' buffer
+ * @param[in]    crcd_data   Source buffer containing combined payload and CRCs
+ *
+ * @return       true if all CRCs are valid
+ * @return       false if at least one CRC is invalid
+ */
+static inline bool _cpy_check_crc(uint8_t *data, size_t len, uint8_t *crcd_data)
+{
+    for (size_t elem = 0; elem < len / 2; elem++) {
+        int idx = (elem << 1);
+        data[idx] = crcd_data[idx + elem];
+        data[idx + 1] = crcd_data[idx + elem + 1];
+
+        if (crc8(&data[idx], 2, SPS30_CRC_POLY, SPS30_CRC_INIT) !=
+            (crcd_data[idx + elem + 2])) {
+            return false;
+        }
+    }
+    return true;
+}
+
+/**
+ * @brief    Communicates with an SPS30 device by reading or writing data.
+ *
+ * @details  This performs all three data transfer types supported by SPS30.
+ *           (1) `Set Pointer`: writes a 16 bit pointer address to the device
+ *           (2) `Set Pointer & Read Data`: (1) followed by separate data-read
+ *           (3) `Set Pointer & Write Data`: (1) combined with a data-write
+ *
+ * @pre          sizeof(crcd_data) must be equal to 1.5 * len
+ *
+ * @param[in]      dev        Pointer to SPS30 device handle
+ * @param[in]      ptr_addr   16 bit pointer address used as command
+ * @param[in/out]  data       Pre-allocated memory pointing to either the data
+ *                            that will be sent to the device or to memory that
+ *                            will hold the response. For type (1) transfers
+ *                            this parameter will be ignored.
+ * @param[in]      len        Length of `data` buffer, set to 0 for type (1)
+ * @param[in]      read       set to true for reading or false for writing
+ *
+ * @return         SPS30_OK           if everyting went fine
+ * @return         -SPS30_CRC_ERROR   if the CRC check failed
+ * @return         -SPS30_I2C_ERROR   if the I2C communication failed
+ */
+static int _rx_tx_data(const sps30_t *dev, uint16_t ptr_addr,
+                       uint8_t *data, size_t len, bool read)
+{
+    int res = -SPS30_I2C_ERROR;
+    unsigned retr = SPS30_ERROR_RETRY;
+
+    if (i2c_acquire(dev->p.i2c_dev) != 0) {
+        LOG_ERROR("could not acquire I2C bus %d\n", dev->p.i2c_dev);
+        return -SPS30_I2C_ERROR;
+    }
+
+    while (res != 0 && retr--) {
```

Comment:
from the logic this should rather be a `do ... while ()` loop, to only decrease `retr` when failed before.

Reply:
done